File: pubmed-literature-mining/scripts/file_storage.py
#!/usr/bin/env python3
"""
File-Based Storage System
Replaces Xata with free file-based storage using JSON files in the repository.
100% free, version-controlled, and perfect for GitHub Actions.
"""

import os
import json
import logging
from typing import Dict, List, Optional
from datetime import datetime
from pathlib import Path

logger = logging.getLogger(__name__)


class FileStorage:
    """File-based storage using JSON files - 100% free alternative to Xata"""
    
    def __init__(self, data_dir: str = 'data/articles'):
        """
        Initialize file storage
        
        Args:
            data_dir: Directory to store article JSON files
        """
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(parents=True, exist_ok=True)
        
        # Index file for fast lookups
        self.index_file = self.data_dir / 'index.json'
        self._load_index()
    
    def _load_index(self):
        """Load or create index file"""
        if self.index_file.exists():
            try:
                with open(self.index_file, 'r') as f:
                    self.index = json.load(f)
            except Exception as e:
                logger.warning(f"Error loading index, creating new one: {e}")
                self.index = {}
        else:
            self.index = {}
    
    def _save_index(self):
        """Save index file"""
        try:
            with open(self.index_file, 'w') as f:
                json.dump(self.index, f, indent=2)
        except Exception as e:
            logger.error(f"Error saving index: {e}")
    
    def _get_article_file(self, pmid: str) -> Path:
        """Get file path for an article"""
        # Store in subdirectories to avoid too many files in one directory
        # Use first 2 digits of PMID for directory structure
        subdir = pmid[:2] if len(pmid) >= 2 else '00'
        article_dir = self.data_dir / subdir
        article_dir.mkdir(exist_ok=True)
        return article_dir / f"{pmid}.json"
    
    def get_article_by_pmid(self, pmid: str) -> Optional[Dict]:
        """Get article by PMID"""
        article_file = self._get_article_file(pmid)
        if not article_file.exists():
            return None
        
        try:
            with open(article_file, 'r') as f:
                return json.load(f)
        except Exception as e:
            logger.error(f"Error reading article {pmid}: {e}")
            return None
    
    def insert_article(self, article_data: Dict) -> bool:
        """
        Insert or update article
        
        Args:
            article_data: Dictionary with article fields
            
        Returns:
            True if successful
        """
        if 'pmid' not in article_data:
            logger.error("PMID is required")
            return False
        
        pmid = article_data['pmid']
        article_file = self._get_article_file(pmid)
        
        try:
            # Load existing article if it exists
            existing = self.get_article_by_pmid(pmid)
            if existing:
                # Merge with existing data (update)
                article_data = {**existing, **article_data}
                article_data['updated_at'] = datetime.now().isoformat()
            else:
                # New article
                article_data['created_at'] = datetime.now().isoformat()
                article_data['updated_at'] = datetime.now().isoformat()
            
            # Save article
            with open(article_file, 'w') as f:
                json.dump(article_data, f, indent=2, ensure_ascii=False)
            
            # Update index
            self.index[pmid] = {
                'pmid': pmid,
                'title': article_data.get('title', '')[:100],  # Truncate for index
                'relevance_score': article_data.get('relevance_score', 0),
                'access_type': article_data.get('access_type', 'unknown'),
                'journal': article_data.get('journal', ''),
                'updated_at': article_data.get('updated_at', '')
            }
            self._save_index()
            
            return True
        except Exception as e:
            logger.error(f"Error saving article {pmid}: {e}")
            return False
    
    def query_articles(self, filter_func=None, limit: int = 100) -> List[Dict]:
        """
        Query articles with optional filter function
        
        Args:
            filter_func: Function that takes article dict and returns bool
            limit: Maximum number of results
            
        Returns:
            List of article dictionaries
        """
        articles = []
        
        # Load all articles from index
        for pmid, index_data in self.index.items():
            if len(articles) >= limit:
                break
            
            article = self.get_article_by_pmid(pmid)
            if not article:
                continue
            
            # Apply filter if provided
            if filter_func is None or filter_func(article):
                articles.append(article)
        
        return articles
    
    def get_high_relevance_articles(self, threshold: int = 70) -> List[Dict]:
        """Get articles with relevance score >= threshold"""
        def filter_func(article):
            return article.get('relevance_score', 0) >= threshold
        return self.query_articles(filter_func=filter_func)
    
    def get_paywalled_articles(self, threshold: int = 70) -> List[Dict]:
        """Get paywalled articles with relevance >= threshold"""
        def filter_func(article):
            return (article.get('access_type') == 'paywalled' and 
                   article.get('relevance_score', 0) >= threshold)
        return self.query_articles(filter_func=filter_func)
    
    def count_paywalled_articles(self) -> int:
        """Count paywalled articles"""
        def filter_func(article):
            return article.get('access_type') == 'paywalled'
        return len(self.query_articles(filter_func=filter_func, limit=10000))
    
    def count_predictive_factors(self) -> int:
        """Count total number of predictive factors extracted"""
        total = 0
        for pmid in self.index.keys():
            article = self.get_article_by_pmid(pmid)
            if article:
                factors = article.get('predictive_factors', [])
                if isinstance(factors, list):
                    total += len(factors)
        return total
    
    def get_all_articles(self) -> List[Dict]:
        """Get all articles (for analysis)"""
        return self.query_articles(limit=10000)
    
    def search_articles(self, search_term: str, limit: int = 100) -> List[Dict]:
        """Search articles by title or abstract"""
        search_lower = search_term.lower()
        def filter_func(article):
            title = article.get('title', '').lower()
            abstract = article.get('abstract', '').lower()
            return search_lower in title or search_lower in abstract
        return self.query_articles(filter_func=filter_func, limit=limit)

File: scripts/file-storage.js
/**
 * File-Based Storage System for Node.js
 * Replaces Xata with free file-based storage using JSON files
 * 100% free, version-controlled, perfect for GitHub Actions
 */

const fs = require("fs");
const path = require("path");

class FileStorage {
  constructor(dataDir = "data") {
    this.dataDir = path.join(__dirname, "..", dataDir);
    this.ensureDirectory(this.dataDir);
  }

  ensureDirectory(dir) {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  // Generic table storage
  getTablePath(tableName) {
    const tableDir = path.join(this.dataDir, tableName);
    this.ensureDirectory(tableDir);
    return tableDir;
  }

  getRecordPath(tableName, id) {
    const tableDir = this.getTablePath(tableName);
    // Use subdirectories to avoid too many files in one directory
    const subdir = id.substring(0, 2) || "00";
    const subdirPath = path.join(tableDir, subdir);
    this.ensureDirectory(subdirPath);
    return path.join(subdirPath, `${id}.json`);
  }

  getIndexPath(tableName) {
    return path.join(this.getTablePath(tableName), "index.json");
  }

  loadIndex(tableName) {
    const indexPath = this.getIndexPath(tableName);
    if (fs.existsSync(indexPath)) {
      try {
        return JSON.parse(fs.readFileSync(indexPath, "utf8"));
      } catch (e) {
        return {};
      }
    }
    return {};
  }

  saveIndex(tableName, index) {
    const indexPath = this.getIndexPath(tableName);
    fs.writeFileSync(indexPath, JSON.stringify(index, null, 2), "utf8");
  }

  async get(tableName, id) {
    const recordPath = this.getRecordPath(tableName, id);
    if (!fs.existsSync(recordPath)) {
      return null;
    }
    try {
      return JSON.parse(fs.readFileSync(recordPath, "utf8"));
    } catch (e) {
      return null;
    }
  }

  async create(tableName, record) {
    if (!record.id) {
      throw new Error("Record must have an 'id' field");
    }

    const recordPath = this.getRecordPath(tableName, record.id);
    const now = new Date().toISOString();

    // Load existing if it exists (for update)
    const existing = await this.get(tableName, record.id);
    if (existing) {
      // Update existing
      const updated = {
        ...existing,
        ...record,
        updated_at: now,
      };
      fs.writeFileSync(
        recordPath,
        JSON.stringify(updated, null, 2),
        "utf8"
      );

      // Update index
      const index = this.loadIndex(tableName);
      index[record.id] = {
        id: record.id,
        title: updated.title || updated.name || "",
        updated_at: now,
      };
      this.saveIndex(tableName, index);

      return updated;
    } else {
      // Create new
      const newRecord = {
        ...record,
        created_at: now,
        updated_at: now,
      };
      fs.writeFileSync(
        recordPath,
        JSON.stringify(newRecord, null, 2),
        "utf8"
      );

      // Update index
      const index = this.loadIndex(tableName);
      index[record.id] = {
        id: record.id,
        title: newRecord.title || newRecord.name || "",
        created_at: now,
        updated_at: now,
      };
      this.saveIndex(tableName, index);

      return newRecord;
    }
  }

  async update(tableName, id, updates) {
    const existing = await this.get(tableName, id);
    if (!existing) {
      throw new Error(`Record ${id} not found in table ${tableName}`);
    }

    const updated = {
      ...existing,
      ...updates,
      updated_at: new Date().toISOString(),
    };

    const recordPath = this.getRecordPath(tableName, id);
    fs.writeFileSync(recordPath, JSON.stringify(updated, null, 2), "utf8");

    // Update index
    const index = this.loadIndex(tableName);
    if (index[id]) {
      index[id].updated_at = updated.updated_at;
      if (updated.title) index[id].title = updated.title;
      if (updated.name) index[id].title = updated.name;
    }
    this.saveIndex(tableName, index);

    return updated;
  }

  async query(tableName, filterFunc = null, limit = 1000) {
    const index = this.loadIndex(tableName);
    const results = [];
    let count = 0;

    for (const id in index) {
      if (count >= limit) break;

      const record = await this.get(tableName, id);
      if (!record) continue;

      if (!filterFunc || filterFunc(record)) {
        results.push(record);
        count++;
      }
    }

    return results;
  }
}

// Xata-compatible API wrapper
class XataCompatibleClient {
  constructor() {
    this.storage = new FileStorage();
  }

  get db() {
    return {
      papers: new TableWrapper(this.storage, "papers"),
      lincs_data: new TableWrapper(this.storage, "lincs_data"),
    };
  }
}

class TableWrapper {
  constructor(storage, tableName) {
    this.storage = storage;
    this.tableName = tableName;
  }

  async get(id) {
    const record = await this.storage.get(this.tableName, id);
    return record ? { id: record.id, ...record } : null;
  }

  async create(record) {
    if (!record.id && !record.pmid) {
      throw new Error("Record must have 'id' or 'pmid' field");
    }
    const id = record.id || record.pmid || record.entrez_id;
    return await this.storage.create(this.tableName, { ...record, id });
  }

  async update(id, updates) {
    return await this.storage.update(this.tableName, id, updates);
  }

  async filter(filterObj) {
    // Support filtering by field
    return {
      getFirst: async () => {
        const allRecords = await this.storage.query(this.tableName);
        // Simple filter: find first record matching filterObj
        for (const record of allRecords) {
          let matches = true;
          for (const key in filterObj) {
            if (record[key] !== filterObj[key]) {
              matches = false;
              break;
            }
          }
          if (matches) {
            return { id: record.id, ...record };
          }
        }
        return null;
      },
      getMany: async () => {
        const allRecords = await this.storage.query(this.tableName);
        // Filter records
        return allRecords.filter((record) => {
          for (const key in filterObj) {
            if (record[key] !== filterObj[key]) {
              return false;
            }
          }
          return true;
        });
      },
    };
  }
}

// Export Xata-compatible client
function getXataClient() {
  // Always use file storage (no API key needed)
  return new XataCompatibleClient();
}

module.exports = { FileStorage, getXataClient, XataCompatibleClient };

